examples/tutorials/core_api_mnist/model_def_adaptive.py

```diff
@@ -0,0 +1,250 @@
+# Participate in the hyperparameter search offered by the Determined master
+
+from __future__ import print_function
+import argparse
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+from torchvision import datasets, transforms
+from torch.optim.lr_scheduler import StepLR
+import torch
+import determined as det
+import pathlib
+
+
+class Net(nn.Module):
+    # NEW: add hparams to __init__
+    def __init__(self, hparams):
+        # NEW: soft code hyperparameters based on config
+        super(Net, self).__init__()
+        self.conv1 = nn.Conv2d(1, hparams["n_filters1"], 3, 1)
+        self.conv2 = nn.Conv2d(hparams["n_filters1"], hparams["n_filters2"], 3, 1)
+        self.dropout1 = nn.Dropout(hparams["dropout1"])
+        self.dropout2 = nn.Dropout(hparams["dropout2"])
+        self.fc1 = nn.Linear(144 * hparams["n_filters2"], 128)
+        self.fc2 = nn.Linear(128, 10)
+
+    def forward(self, x):
+        x = self.conv1(x)
+        x = F.relu(x)
+        x = self.conv2(x)
+        x = F.relu(x)
+        x = F.max_pool2d(x, 2)
+        x = self.dropout1(x)
+        x = torch.flatten(x, 1)
+        x = self.fc1(x)
+        x = F.relu(x)
+        x = self.dropout2(x)
+        x = self.fc2(x)
+        output = F.log_softmax(x, dim=1)
+        return output
+
+
+# NEW: Modified function header to include op for reporting training progress to master
+def train(args, model, device, train_loader, optimizer, core_context, epoch, op):
+    model.train()
+    for batch_idx, (data, target) in enumerate(train_loader):
+        data, target = data.to(device), target.to(device)
+        optimizer.zero_grad()
+        output = model(data)
+        loss = F.nll_loss(output, target)
+        loss.backward()
+        optimizer.step()
+        if (batch_idx + 1) % args.log_interval == 0:
+
+            print(
+                "Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}".format(
+                    epoch,
+                    (batch_idx) * len(data),
+                    len(train_loader.dataset),
+                    100.0 * (batch_idx) / len(train_loader),
+                    loss.item(),
+                )
+            )
+
+            # NEW: report progress once in a while
+            op.report_progress(epoch)
```

Comment:
Probably you should either report partial epochs (that is, `epoch + batch_idx / len(train_loader)), or only report the epoch on epoch boundaries.

Multiple progress reports with the same value just slows down training, with no visible effect on the webui's progress reporting.